src/dfmt/formatter.d

```diff
@@ -121,8 +123,23 @@ struct TokenFormatter(OutputRange)
                 this.eolString = "\n";
             else if (eol == eol.crlf)
                 this.eolString = "\r\n";
-            else if (eol == eol.unspecified)
+            else if (eol == eol._unspecified)
                 assert(false, "config.end_of_line was unspecified");
+            else
+            {
+                assert (eol == eol._default); // Same as input.
+                // Intentional wraparound, -1 turns into uint.max when not found:
+                const firstCR = cast(uint) rawSource.countUntil("\r");
+                if (firstCR < cast(uint) rawSource.countUntil("\n"))
+                {
+                    if (firstCR == rawSource.countUntil("\r\n"))
+                        this.eolString = "\r\n";
+                    else
+                        this.eolString = "\r";
+                }
+                else
+                    this.eolString = "\n";
```

Comment:
isolate this logic into its own function pls, constructor might not be a good place for application logic.